File: 407_plus-one/plus-one.cpp
/*
@Copyright:LintCode
@Author:   shaman
@Problem:  http://www.lintcode.com/problem/plus-one
@Language: C++
@Datetime: 16-04-04 16:54
*/

class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        digits.back() += 1;
        for (int i = digits.size() - 1; i > 0; i--) {
            if (digits[i] >= 10) {
                digits[i] -= 10;
                digits[i-1] += 1;
            } else {
                break;
            }
        }
        
        if (digits[0] >= 10) {
            digits[0] = 1;
            fill(digits.begin() + 1, digits.end(), 0);
            digits.push_back(0);
        }
        
        return digits;
    }
};

File: 407_plus-one/plus-one.java
/*
@Copyright:LintCode
@Author:   shaman
@Problem:  http://www.lintcode.com/problem/plus-one
@Language: Java
@Datetime: 16-04-04 16:57
*/

public class Solution {
    public int[] plusOne(int[] digits) {
        return plusOne(digits, digits.length - 1);
    }

    private int[] plusOne(int[] digits, int index) {
        int digit = digits[index] + 1;
        if (digit < 10) {
            digits[index] = digit;
            return digits;
        }
        digits[index] = 0;
        if (index == 0) {
            int[] newDigits = new int[digits.length + 1];
            newDigits[0] = 1;
            System.arraycopy(digits, 0, newDigits, 1, digits.length);
            return newDigits;
        }
        return plusOne(digits, index - 1);
    }
}